Automattic/jetpack
# Stats: switch to Core WP methods to add the tracking pixel onto pages

Fixes #29560

## Proposed changes:

This is a different approach, following our discussion in https://github.com/Automattic/jetpack/pull/29583#discussion_r1151065999.

Instead of working around how we currently add the tracking pixel onto pages by manually creating and outputting the pixel and its script into `wp_footer`, let's use core functions and filters (`wp_enqueue_script`).

To make this possible, I had to make changes to the Assets package as well, since the method used to add `defer` wasn't previously easily accessible from elsewhere.

With this change, one could use the `script_loader_tag` filter to change how the Stats tracking pixel is outputted:

```php
/**
 * Add Stats script attributes a different way.
 *
 * @param string $tag The <script> tag for the enqueued script.
 * @param string $handle The script's registered handle.
 */
function jeherve_add_stats_script_attributes_2( $tag, $handle ) {
	if ( 'jetpack-stats' !== $handle ) {
		return $tag;
	}

	return preg_replace( '/<script( [^>]*)? src=/i', '<script type="text/plain"$1 src=', $tag );
}
add_filter( 'script_loader_tag', 'jeherve_add_stats_script_attributes_2', 10, 2 );
```

### Other information:


## Jetpack product discussion

* https://github.com/Automattic/jetpack/pull/29583#discussion_r1151065999
* p1680105234226939-slack-CBG1CP4EN

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

1. Start with a site that's connected to WordPress.com, with the stats module active.
2. Open the site in an incognito window, and look for the `<script src='https://stats.wp.com/` script tag.
    * It should include just the `defer` attribute.
3. Now add the code snippet above to a functionality plugin.
    * Ensure the new attribute is correctly added.
4. Do the tests pass?

I would also recommend to ensure the stats pixel is still added to pages when using the AMP plugin and its different settings.
